File: apps/api/src/integrations/_mocks/transport.ts
import type { HttpRequest, HttpResponse, HttpTransport } from '../_http/client.js';
import { getMockStore } from './store.js';
import { handleQbo } from './qbo.js';
import { handleHarvest } from './harvest.js';
import { handleDropboxsign } from './dropboxsign.js';
import { handleAirtable } from './airtable.js';
import { handleGmail } from './gmail.js';

// Route HTTP requests to per-source mock handlers based on the URL host.
// Each handler returns a response object (status + body). Failures injected
// via the FailureRegistry preempt the handler.

export class MockTransport implements HttpTransport {
  async request(req: HttpRequest): Promise<HttpResponse> {
    const url = new URL(req.url);
    const host = url.host;
    const path = url.pathname;

    // Allow tests to inject the next failure for a key. Mock handlers
    // call `consumeFailureFor(<key>)` at their entry; if a failure is
    // queued, they short-circuit with that response.
    const store = getMockStore();
    void store;

    const route = (handler: () => Promise<HttpResponse>): Promise<HttpResponse> => handler();

    if (host.endsWith('intuit.com') || host.includes('quickbooks')) {
      return route(() => handleQbo(req, path));
    }
    if (host.includes('harvestapp.com') || host.includes('harvest.com')) {
      return route(() => handleHarvest(req, path));
    }
    if (host.includes('dropboxsign.com') || host.includes('hellosign.com')) {
      return route(() => handleDropboxsign(req, path));
    }
    if (host.includes('airtable.com')) {
      return route(() => handleAirtable(req, path));
    }
    if (host.includes('googleapis.com') || host.includes('gmail.com')) {
      return route(() => handleGmail(req, path));
    }

    return {
      status: 404,
      headers: {},
      body: { error: `unmocked host: ${host}` },
    };
  }
}

export function jsonResponse(status: number, body: unknown): HttpResponse {
  return {
    status,
    headers: { 'content-type': 'application/json' },
    body,
  };
}

export function consumeFailureFor(key: string): HttpResponse | null {
  const store = getMockStore();
  const f = store.failures.consume(key);
  if (!f) return null;
  return {
    status: f.status,
    headers: { 'content-type': 'application/json' },
    body: f.body ?? { error: `injected failure (${f.status})` },
  };
}

File: apps/api_py/src/linkbook/integrations/mocks/transport.py
"""Mock HTTP transport. Routes requests to per-source handlers by hostname.

Per-source live escape hatch: if a hostname maps to a source listed in
INTEGRATION_LIVE_SOURCES, this transport delegates to the real httpx
transport instead of the mock handler. Lets you take one source live
(e.g. harvest) while everything else stays mocked.
"""

from __future__ import annotations

from typing import Any

from ...config import live_sources, load_config, source_for_host
from ..http import FetchTransport, HttpRequest, HttpResponse, host_of, set_transport
from .store import get_mock_store


def json_response(status: int, body: Any) -> HttpResponse:
    return HttpResponse(status=status, headers={"content-type": "application/json"}, body=body)


def consume_failure_for(key: str) -> HttpResponse | None:
    store = get_mock_store()
    spec = store.failures.consume(key)
    if spec is None:
        return None
    return HttpResponse(
        status=spec.status,
        headers={"content-type": "application/json"},
        body=spec.body if spec.body is not None else {"error": f"injected failure ({spec.status})"},
    )


class MockTransport:
    """Routes requests to per-source handlers by hostname.

    Sources whose name is in INTEGRATION_LIVE_SOURCES skip mocking and
    pass through to a real httpx transport.
    """

    def __init__(self) -> None:
        # Built lazily so a config reload during the process can be picked
        # up by a fresh MockTransport instance, but per-request lookup is
        # O(1) against the cached set.
        cfg = load_config()
        self._live: set[str] = live_sources(cfg)
        # Real transport reused across all live-source requests.
        self._real: FetchTransport | None = FetchTransport() if self._live else None

    async def request(self, req: HttpRequest) -> HttpResponse:
        host = host_of(req.url)
        path = req.url[req.url.index(host) + len(host) :].split("?", 1)[0] if host else req.url

        if not host:
            return json_response(404, {"error": f"unmocked url: {req.url}"})

        # Per-source live override — bypass mocks for this request.
        source = source_for_host(host)
        if source and source in self._live and self._real is not None:
            return await self._real.request(req)

        if "intuit.com" in host or "quickbooks" in host:
            from . import qbo

            return await qbo.handle(req, path)
        if "harvestapp.com" in host or "harvest.com" in host:
            from . import harvest

            return await harvest.handle(req, path)
        if "dropboxsign.com" in host or "hellosign.com" in host:
            from . import dropboxsign

            return await dropboxsign.handle(req, path)
        if "airtable.com" in host:
            from . import airtable

            return await airtable.handle(req, path)
        if "googleapis.com" in host or "gmail.com" in host:
            from . import gmail

            return await gmail.handle(req, path)

        return json_response(404, {"error": f"unmocked host: {host}"})


def install_mock_transport() -> None:
    set_transport(MockTransport())
